doc/design.md

```diff
@@ -0,0 +1,1268 @@
+More code is likely to require templating for the precision (`double` and `single`).
+However, the development cost for this library itself would be rather minor, so there is very little risk involved in supporting it from the beginning.
+
+
+### <a name="design-details-slicing"></a>Slicing
+
+We would like to support slicing as in `numpy` and `xarray`, e.g., `array[2:4,10:20, :]` in the case of slicing `Variable` (or the equivalent with named dimensions in the case of slicing `Dataset`, as supported in `xarray`).
+In both `numpy` and `xarray` the returned object is a view, i.e., data is not extracted.
+
+The ownership and copying discussion in the section [Copy-on-write mechanism](#design-details-copy-on-write-mechanism) is related to such slice-views and implies:
+- Any iterator obtained from a view is invalidated if any non-`const` method on *any other view* to the same `Variable` is called.
+  - The view itself can stay valid since it references the variable, not its data.
+- Contrary to the `numpy` behavior, if the owning `Variable` goes out of scope all views should probably become invalid.
+  To support views that stay valid we would need to wrap all variables in `Dataset` into a `shared_ptr` and the benefit does not seem to justify this additional complexity at this point.
+
+
+## <a name="implementation"></a>Implementation
+
+
+### <a name="implementation-goals-and-non-goals"></a>Goals and non-goals
+
+##### Goals in this implementation phase
+
+1. Provide a simple, lightweight, functional and well-tested library that can replace and enhance most of the workspaces types existing in Mantid.
+1. Reach a point where this library (in combination with some supporting libraries) is useful in practice.
+   We define useful as:
+   - Complete low-level functionality in C++ and Python.
+   - `numpy` interoperability.
+   - Visualization using a slice viewer.
+   - Visualization using a multi-dimensional table viewer.
+   - Instrument visualization using the Mantid instrument viewer.
+   - Can leverage existing Mantid code such as `DataHandling::LoadEventNexus` by a subsequent conversion into a `Dataset`.
+1. Demonstrate usefulness on at least two real examples, ideally bringing them to a production ready stage.
+   Candidates are:
+   - Constant-wavelength (reactor) workflows, which currently use `API::MatrixWorkspace` with length-1 histograms as a workaround.
+   - Imaging workflows, which would benefit from presenting data as a stack of images rather than a list of histograms, i.e., the dimension order would be the opposite of what we are used to in `API::MatrixWorkspace`.
+   - Workflow with parameter scan or polarization analysis to demonstrate new capabilities of extra dimensions and/or multiple data variables.
+1. If applicable, demonstrate performance gain over existing Mantid workflow.
+
+##### Non-goals
+
+Even beyond this implementation phase, we have the following non-goals:
+
+1. It is a non-goal to provide a drop-in replacement for
```

Comment:
I understand the challenge here. My fear here is confusion for users at the Python level when it won't be obvious whether to use the `dataset` api or the traditional one. Would we consider a `v2` algorithm python api where the new versions were one that use dataset alone?

Reply:
For now all the functionality described here is functional, i.e., there as a small and well defined set of functions that can be apply to datasets. That is, at this level we do not have the traditional algorithm inheriting `API::Algorithm`. As you say, ideally this would be a `v2` Python API that is useful on its own. I do not have very clear ideas for this now, and I am not sure we can decide at this phase of design/implementation?

I do not know if we eventually want to wrap all functionality in wrapper algorithm inheriting `API::Algorithm`?